src/mame/drivers/smartstart.cpp

```diff
@@ -0,0 +1,310 @@
+// license:GPL-2.0+
+// copyright-holders:Felipe Sanches
+/*
+    PenseBem (TecToy) / SmartStart (VTech)
+    driver by Felipe Correa da Silva Sanches <juca@members.fsf.org>
+
+---------------------------------------------------------------------
+
+    The 2017 edition of TecToy's PenseBem has a 2x4 programming
+    pin-header at position CN4:
+
+    CN4 - ATMEGA
+      1 -  4 VCC
+      2 - 15 MOSI
+      3 - 31 TXD
+      4 - 16 MISO
+      5 - 30 RXD
+      6 - 17 SCK
+      7 -  5 GND
+      8 - 29 RESET
+
+    R34 is the pull-up resistor for the RESET signal
+
+---------------------------------------------------------------------
+
+    Changelog:
+
+    2020 OCT 27 [Felipe Sanches]:
+        * Fixed keyboard inputs, display & buzzer.
+
+    2017 OCT 07 [Felipe Sanches]:
+        * Initial driver skeleton
+
+---------------------------------------------------------------------
+
+== Notes about the hardware: ==
+=== Select keypad rows: ===
+keypad pin 1 - PORT_B5
+keypad pin 2 - PORT_B3
+keypad pin 12 - PORT_B2
+keypad pin 13 - PORT_B4
+
+=== Read keypad Columns: ===
+keypad pin 3 - PORT_E0
+keypad pin 4 - PORT_D2
+keypad pin 5 - PORT_E1
+keypad pin 6 - PORT_D7
+keypad pin 7 - PORT_D6
+keypad pin 8 - PORT_D5
+keypad pin 9 - PORT_D4
+keypad pin 10 - PORT_D3
+
+=== Display digits: ===
+digit_0 - PORT_C5
+digit_1 - PORT_C4
+digit_2 - PORT_C3
+digit_3 - PORT_C2
+digit_4 - PORT_C1
+digit_5 - PORT_C0
+digit_6 - PORT_E3
+digit_7 - PORT_E2
+
+=== Display segments: ===
+In parentheses are the resistor reference numbers on
+the PCB for each of these signals.
+
+seg_7 (R11) - PORT_D7
+seg_6 (R12) - PORT_D6
+seg_5 (R17) - PORT_D5
+seg_4 (R13) - PORT_D4
+seg_3 (R14) - PORT_D3
+seg_2 (R15) - PORT_D2
+seg_1 (R18) - PORT_E1
+seg_0 (R16) - PORT_E0
+
+=== Piezo buzzer: ===
+Port B, bit 1
+
+--------------------------------------------------------------------- */
+
+#include "emu.h"
+#include "cpu/avr8/avr8.h"
+#include "sound/dac.h"
+#include "rendlay.h"
+#include "screen.h"
+#include "speaker.h"
+
+class pensebem2017_state : public driver_device
+{
+public:
+	pensebem2017_state(const machine_config &mconfig, device_type type, const char *tag)
+		: driver_device(mconfig, type, tag),
+		m_maincpu(*this, "maincpu"),
+		m_dac(*this, "dac"),
+		m_out_x(*this, "%u.%u", 0U, 0U),
+		m_keyb_rows(*this, "ROW%u", 0U)
+	{
+	}
+	uint8_t m_port_b;
+	uint8_t m_port_c;
+	uint8_t m_port_d;
+	uint8_t m_port_e;
+
+	required_device<avr8_device> m_maincpu;
+	required_device<dac_bit_interface> m_dac;
+	output_finder<8, 8> m_out_x;
+	required_ioport_array<4> m_keyb_rows;
+
+	uint8_t port_r(offs_t offset);
+	void port_w(offs_t offset, uint8_t value);
+	void prg_map(address_map &map);
+	void data_map(address_map &map);
+	void io_map(address_map &map);
+	void pensebem2017(machine_config &config);
+	void update_display();
```

Comment:
Most of these members can be private.  The constructor and machine configuration function need to be public, but the other members don’t need to be.